include/envoy/http/async_client.h

```diff
@@ -62,6 +128,16 @@ class AsyncClient {
    */
   virtual Request* send(MessagePtr&& request, Callbacks& callbacks,
                         const Optional<std::chrono::milliseconds>& timeout) PURE;
+
+  /**
+   * Start an HTTP stream asynchronously
+   * @param callbacks the callbacks to be notified of stream status.
+   * @return a stream handle or nullptr if no stream could be started. NOTE: In this case
+   *         onResetStream() has already been called inline. The client owns the stream and
+   *         the handle can be used to send more messages or close the stream.
+   */
+  virtual Stream* start(StreamCallbacks& callbacks,
+                        const Optional<std::chrono::milliseconds>& timeout) PURE;
```

Comment:
When timeout, it will call StreamCallbacks.onHeader with :status 504 and onData, the header and data is coming from underlying filters.

See this test:
https://github.com/lyft/envoy/pull/353/files#diff-918cf37446b66e452d1114b88b1e61c4R578